pystiche_papers/sanakoyeu_et_al_2018/_data.py

```diff
@@ -85,29 +93,171 @@ def _properties(self) -> Dict[str, Any]:
+        return dict(
+            src=src,
+            dst=dst,
+            interpolation=torch.tensor(kornia.Resample.get(self.interpolation).value),
+            align_corners=torch.tensor(self.align_corners),
+        )
+
+    @staticmethod
+    def generate_anchors(
+        batch_size: int,
+        image_size: Tuple[int, int],
+        crop_size: Tuple[int, int],
+        same_on_batch: bool,
+    ) -> torch.Tensor:
+        def generate_single_dim_anchor(
+            batch_size: int, image_length: int, crop_length: int, same_on_batch: bool
+        ) -> torch.Tensor:
+            diff = image_length - crop_length
+            if diff <= 0:
+                return torch.zeros((batch_size,), dtype=torch.int)
+            else:
+                return _adapted_uniform_int((batch_size,), 0, diff, same_on_batch)
+
+        single_dim_anchors = [
+            generate_single_dim_anchor(
+                batch_size, image_length, crop_length, same_on_batch
+            )
+            for image_length, crop_length in zip(image_size, crop_size)
+        ]
+        return torch.stack(single_dim_anchors, dim=1,).unsqueeze(1).repeat(1, 4, 1)
+
+    @staticmethod
+    def generate_vertices_from_size(
+        batch_size: int, size: Tuple[int, int]
+    ) -> torch.Tensor:
+        height, width = size
+        return (
+            torch.tensor(
+                ((0, 0), (width - 1, 0), (width - 1, height - 1), (0, height - 1),),
+                dtype=torch.int,
+            )
+            .unsqueeze(0)
+            .repeat(batch_size, 1, 1)
+        )
+
+    @staticmethod
+    def clamp_vertices_to_size(
+        vertices: torch.Tensor, size: Tuple[int, int]
+    ) -> torch.Tensor:
+        horz, vert = vertices.split(1, dim=2)
+        height, width = size
+        return torch.cat(
+            (torch.clamp(horz, 0, width - 1), torch.clamp(vert, 0, height - 1),), dim=2,
+        )
+
+    def compute_transformation(
+        self, input: torch.Tensor, params: Dict[str, torch.Tensor]
+    ) -> torch.Tensor:
+        return cast(torch.Tensor, compute_crop_transformation(input, params))
+
+    def apply_transform(
+        self, input: torch.Tensor, params: Dict[str, torch.Tensor]
+    ) -> torch.Tensor:
+        return cast(torch.Tensor, apply_crop(input, params))
+
+    def _properties(self) -> Dict[str, Any]:
+        dct = super()._properties()
+        dct["size"] = self.size
+        if self.interpolation != "bilinear":
+            dct["interpolation"] = self.interpolation
+        if self.same_on_batch:
+            dct["same_on_batch"] = True
+        if self.align_corners:
+            dct["align_corners"] = True
+        return dct
+
+
+def image_transform(
+    impl_params: bool = True, edge_size: int = 768, train: bool = False
 ) -> nn.Sequential:
     transforms_: List[nn.Module] = [
-        ClampSize(),
-        transforms.ValidRandomCrop(image_size),
-        OptionalGrayscaleToFakegrayscale(),
```

Comment:
@jbueltemeier Do you remember why we included this? I can't find any evidence of this in their implementation? If not, I think it is OK to remove it.

Reply:
This was only introduced to get the `image_size `of 768x768, because there was no `augmentation `before. So you are right that it is no longer needed and can be removed.

Reply:
Ok, I just don't understand why it makes sense here to split the augmentation. 

What we need is if:

1. `impl_params=True and train=True`: a sequence a sequence of `ClampSize -> augmentation`  with internal cropping

2. `impl_params=True and train=False`: Resize the smallest side of the image to the image_size and optionally a transformation back to the original image size. ([resize](https://github.com/pmeier/adaptive-style-transfer/blob/07a3b3fcb2eeed2bf9a22a9de59c0aea7de44181/model.py#L494), [optional back](https://github.com/pmeier/adaptive-style-transfer/blob/07a3b3fcb2eeed2bf9a22a9de59c0aea7de44181/model.py#L506)).

3.  `impl_params=false`: An interpretation of how to handle images that are smaller than the `image_size` in the quote below. So maybe a sequence of `OptionalUpsample -> ValidRandomCrop`.
> trained from scratch on randomly cropped image patches of size 768×768 pix.

Reply:
In the reference implementation the training pipeline goes like this:

1. `ClampSize`
2. `RandomRescale`
3. `DynamicSizePad2d`
    1. `RandomRotation`
    2. `RandomAffine`
4. `RandomCrop`
5. `RandomHSVJitter`
6. `RandomHorizontalFlip`

I've aggregated 2. and 3. to `pre_crop_augmentation` and 5. and 6. to `post_crop_augmentation`.

That leaves us with the following cases

- `impl_params and train`: see pipeline above
- `impl_params and not train`: remove `pre_crop_augmentation` and `post_crop_augmentation` leaving only `ClampSize` and `RandomCrop`
- `not impl_params`: I think (and if I understand you correctly you agree) the transformation that achieves 
    > randomly cropped image patches of size 768×768 pix

    is an optional upsample to 768 pixels on the shorter edge and subsequently a valid random crop.

As is, my implementation does exactly that.

---

> Resize the smallest side of the image to the image_size and optionally a transformation back to the original image size.

This is only for the images that you want to stylize after the training. There is no reason to pass them through these transforms at all. If we want this functionality at all, and I'm not sure about that yet, it should be implemented in `paper.nst()`.

Reply:
Okay, I agree with you on cases one and three. Only the case `impl_params and not train` I do not find in the implementation. This should be the case `is_inference` in their implementation and there only a `resize` is done instead of the `crop`. See [here](https://github.com/CompVis/adaptive-style-transfer/blob/07a3b3fcb2eeed2bf9a22a9de59c0aea7de44181/img_augm.py#L60).  Or am I still thinking wrong?

Reply:
No, you were right. `train=True` should always be the case, since `image_transforms` are only used during training. I've removed it in the recent commits.